src/Microsoft.ML.LightGBM/LightGbmTrainerBase.cs

```diff
@@ -89,6 +90,116 @@ private sealed class CategoricalMetaData
             InitParallelTraining();
         }
 
+        [BestFriend]
+        internal void ExpandMonotoneConstraint(ref Dictionary<string, object> options, int featureCount)
+        {
+            string monotoneArgumentName = "monotone_constraints";
+            if (!options.ContainsKey(monotoneArgumentName))
+                return;
+
+            // If the constraints arguments is not a string array, then return
+            if (!(options[monotoneArgumentName] is string[]))
+                return;
+
+            string[] constraintArguments = (string[])options[monotoneArgumentName];
+            if (constraintArguments == null || constraintArguments.Length == 0)
+            {
+                options.Remove(monotoneArgumentName);
+                return;
+            }
+
+            // Convert the monotone constraints parameter to be consumed by LightGBM.
+            // Format of the constraint is the key word that specifies
+            // the constraint type: pos for positive constraint (which results to
+            // 1 for LightGBM) and neg for negative constraint (which results to -1
+            // for LightGBM).
+            // For example:
+            // pos:0-3,5
+            // neg:6-7
+            // Since the argument is a string array, mulitple ranges can be specified.
+            // In the event of the same index being specified twice, the last one wins.
+            // The end result is a string array of 1s,0s, and -1s that should be equal
+            // to the number of features in the Feature column.
+            int[] constraintArray = new int[featureCount];
+            const string positiveKeyword = "pos";
+            const string negativeKeyword = "neg";
+
+            foreach(var argument in constraintArguments)
+            {
+                // Split by : to get the keyword and range
+                var subArguments = argument.Split(':');
+                if (subArguments.Length > 2)
+                    // Invalid argument, skip to the next argument
+                    continue;
+
+                var keyword = subArguments[0].ToLowerInvariant();
+                int constraint = 0;
+                if (keyword.Equals(positiveKeyword, StringComparison.OrdinalIgnoreCase))
+                    constraint = 1;
+                else if (keyword.Equals(negativeKeyword, StringComparison.OrdinalIgnoreCase))
+                    constraint = -1;
+                else
+                    continue;
```

Comment:
We're also silently ignore the user input here.

Reply:
Thanks Justin, I am now throwing.